### PR TITLE
New NetSocket interface with common BSD/Win implementation

### LINKED ISSUE #20214: Multiplayer client crash when getting a lot of rset calls
**Godot version:**
Godot 3.0.5

**OS/device including version:**
Tested on Windows 10(64 bit), Windows 8.1(32 bit)

**Issue description:**
In a client/server application using Godot's implementation of ENet, the client's window stops responding when it's receiving a lot of networked calls(I tested rset_unreliable and rpc). In my case server is the master of a few hundredth entities(I tested for ~500), each of them sends(on each physics update) rset_unreliable to set Vector2 variable on the slave side.

**Minimal reproduction project:**
[multiplayer_crash.zip](https://github.com/godotengine/godot/files/2201301/multiplayer_crash.zip)
Run two instances of the project, in one of them press host, in the other press join.
Client should hang and stop responding(immediately or after a few seconds).

### PR DESCRIPTION
**Pull request by Faless**
This is long awaited, at least for me.

Unified BSD and Winsock sockets into a single implementation of a generic NetSocket interface.

This is an important ground work for few network improvements:
- Reuse as much code as possible between Posix and Windows. (*done*)
- Provide a single point of implementation for exotic sdks (consoles). (*done*)
- Provide platform agnostic `StreamPeerTCP` and `PacketPeerUDP` in core. (*done*)
- Implement connect for UDP allowing for DTLS implementation. (*todo*)

This allows will also make further network improvements easier to implement, and possibly result in less bug as we have much less code now.

Closes #20214

### CODE REVIEW

**Comment by Faless** [comment]
Fixed windows builds, tested on Linux + builds on:
- OSX (cross compiling)
- Iphone (cross compiling)
- Windows (native and cross compiling)
- UWP (64 bits, native)
- Linux/OpenBSD/FreeBSD
- Android (compiling from Linux)
- HTML5

**Comment by Faless** [comment]
Forced push to remove unused instance variables in PacketPeerUDP

**Comment by punto-** [comment]
Looks good, I'm not sure if we can assume that all low level networking apis will use this model where all the sockets are in the same place, but for now I'll trade that future hypothetical problem for the reduction in code. Great job.

Out of curiosity, what does this do in HTML5?

**Comment by Faless** [comment]
Thanks!

> Out of curiosity, what does this do in HTML5?

Well, just like before. We keep using BSD sockets which EMScripten tries translates to browser primitives. TCP becomes a WebSocket connection with "binary" sub protocol name, UDP is broken.
Those low level classes cannot really be used if Webassembly is a target, since they are not supported by browser themselves.
That's why we ended up needing native WebSocket too and soon:tm:  WebRTC.

**Comment by Faless** [comment]
Rebased after big header update

File: platform/uwp/os_uwp.cpp [hunk 1]
```diff
@@ -28,6 +28,9 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
 /*************************************************************************/
 
+// Must include Winsock before windows.h (included by os_windows.h)
```
Comment (akien-mga) [inline platform/uwp/os_uwp.cpp]: s/os_windows/os_uwp/

**Comment by meld-cp** [comment]
Hi, I think this commit broke the debugger from opening a listening socket when we run a project...  for Win 10 anyway.... I'm happy to do more testing if needed.

I get these errors in the console:

```
ERROR: NetSocketPosix::open: Condition ' _sock == (SOCKET)(~0) ' is true. returned: FAILED
   At: drivers\unix\net_socket_posix.cpp:232
ERROR: TCP_Server::listen: Condition ' err != OK ' is true. returned: ERR_CANT_CREATE
   At: core\io\tcp_server.cpp:56
ERROR: NetSocketPosix::open: Condition ' _sock == (SOCKET)(~0) ' is true. returned: FAILED
   At: drivers\unix\net_socket_posix.cpp:232
ERROR: StreamPeerTCP::connect_to_host: Condition ' err != OK ' is true. returned: FAILED
   At: core\io\stream_peer_tcp.cpp:73
ERROR: ScriptDebuggerRemote::connect_to_host: Remote Debugger: Unable to connect. Status: 0
   At: core\script_debugger_remote.cpp:92
```

**Comment by Faless** [comment]
@meld-cp I confirm the issue, thanks for the report, not sure how this happened, I did tested it at some point, I'll investigate further.

**Comment by Faless** [comment]
I found the issue, silly me, working on a fix now.

**Pull request merged**